bin/echo.c

```diff
@@ -70,6 +70,75 @@ static int wait_for_event(int fd, s2n_blocked_status blocked)
     return S2N_SUCCESS;
 }
 
+int recv_session_ticket(struct s2n_connection *conn, int fd)
+{
+    s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+    uint8_t out[100] = { 0 };
+    ssize_t bytes_read = 0;
+
+    /* s2nc needs to call s2n_recv until it successfully reads some data from the server.
+     * The output buffer is an arbitrary size to catch whatever bytes the server is sending. */
+    do {
+        bytes_read = s2n_recv(conn, &out, sizeof(out), &blocked);
+    } while (bytes_read < 0);
+
+    if (bytes_read > 0) {
+        for (size_t i = 0; i < bytes_read; i++) {
+            fprintf(stdout, "%c", out[i]);
+        }
+    }
+
+    return S2N_SUCCESS;
+}
+
+int early_data_recv(struct s2n_connection *conn)
+{
+    uint32_t max_early_data_size = 0;
+    GUARD_RETURN(s2n_connection_get_max_early_data_size(conn, &max_early_data_size), "Error getting max early data size");
+    if (max_early_data_size == 0) {
+        return S2N_SUCCESS;
+    }
+
+    ssize_t total_data_recv = 0;
+    ssize_t data_recv = 0;
+    bool server_success = 0;
+    s2n_blocked_status blocked = 0;
+    uint8_t *early_data_received = malloc(max_early_data_size);
+
+    do {
+        server_success = (s2n_recv_early_data(conn, early_data_received + total_data_recv,
```

Comment:
If `s2n_recv_early_data` fails without receiving any early data in `early_data_recv`, we will be in an infinite loop here before we timeout, is that intended?

Reply:
Yes that is intended, this function will only succeed if we receive the End of Early data message, otherwise we cannot continue.

Reply:
To be clear: s2n_recv_early_data succeeds if there's no early data to read. Like s2n_negotiate, it succeeds once there's no work left for it to do.